File: go/parsers/abort.go
package parsers

import (
	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-go/serializers"
	"github.com/xconnio/wampproto-serializer-capnproto/go/gen"
)

type Abort struct {
	gen *gen.Abort
	ex  *PayloadExpander
}

func NewAbortFields(g *gen.Abort, payload []byte) messages.AbortFields {
	return &Abort{
		gen: g,
		ex:  &PayloadExpander{payload: payload, serializer: g.PayloadSerializerID()},
	}
}

func (a *Abort) Reason() string {
	reason, _ := a.gen.Reason()
	return reason
}

func (a *Abort) Details() map[string]any {
	return map[string]any{}
}

func (a *Abort) Args() []any {
	return a.ex.Args()
}

func (a *Abort) KwArgs() map[string]any {
	return a.ex.Kwargs()
}

func AbortToCapnproto(m *messages.Abort) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	abort, err := gen.NewRootAbort(seg)
	if err != nil {
		return nil, err
	}

	if err = abort.SetReason(m.Reason()); err != nil {
		return nil, err
	}

	payloadSerializer := selectPayloadSerializer(m.Details())
	abort.SetPayloadSerializerID(payloadSerializer)
	payload, err := serializers.SerializePayload(payloadSerializer, m.Args(), m.KwArgs())
	if err != nil {
		return nil, err
	}

	data, err := msg.MarshalPacked()
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeAbort, data, payload), nil
}

func CapnprotoToAbort(data, payload []byte) (*messages.Abort, error) {
	msg, err := capnp.UnmarshalPacked(data)
	if err != nil {
		return nil, err
	}

	abort, err := gen.ReadRootAbort(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewAbortWithFields(NewAbortFields(&abort, payload)), nil
}

File: python/wamp_msgs_capnp/parsers/abort.py
import os
from typing import Any
from pathlib import Path

import capnp
from wampproto.messages.abort import Abort, IAbortFields
from wampproto.serializers.payload import serialize_payload

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
module_file = os.path.join(root_dir, "schemas", "abort.capnp")
abort_capnp = capnp.load(str(module_file))


class AbortFields(IAbortFields):
    def __init__(self, gen, payload: bytes):
        self._gen = gen
        self._ex = helpers.PayloadExpander(payload, gen.payloadSerializerID)

    @property
    def reason(self) -> str:
        return self._gen.reason

    @property
    def details(self) -> dict[str, Any]:
        return {}

    @property
    def args(self) -> list[Any]:
        return self._ex.args

    @property
    def kwargs(self) -> dict[str, Any]:
        return self._ex.kwargs

    @property
    def payload_serializer_id(self) -> bytes:
        return self._gen.payloadSerializerID

    @property
    def payload(self) -> bytes:
        return self._ex.payload


def abort_to_capnproto(a: Abort) -> bytes:
    abort = abort_capnp.Abort.new_message()
    abort.reason = a.reason
    payload_serializer = helpers.select_payload_serializer(a.details)
    abort.payloadSerializerID = payload_serializer

    payload = serialize_payload(payload_serializer, a.args, a.kwargs)
    packed_data = abort.to_bytes_packed()

    return helpers.prepend_header(Abort.TYPE, packed_data, payload)


def capnproto_to_abort(data: bytes) -> Abort:
    message_data, payload_data = helpers.extract_message(data)
    abort_obj = abort_capnp.Abort.from_bytes_packed(message_data)

    return Abort(AbortFields(abort_obj, payload_data))
